File: lib/models/revenue_ledger.dart
import 'profile.dart';
import 'membership_plan.dart';
import 'gym.dart';

class RevenueLedger {
  final String id;
  final DateTime paidAt;
  final String? planId;
  final String? membershipId;
  final String? gymId;
  final int amountCents;
  final String currency;
  final String source;
  final String entryKind;
  final String paymentMethod;
  final String? notes;
  final String? recordedBy;
  final bool isDeleted;
  final DateTime? deletedAt;
  final String? deletedBy;
  final DateTime createdAt;

  // Relations
  final Profile? recordedByProfile;
  final Profile? clientProfile;
  final MembershipPlan? plan;
  final Gym? gym;
  final Profile? deletedByProfile;

  RevenueLedger({
    required this.id,
    required this.paidAt,
    this.planId,
    this.membershipId,
    this.gymId,
    required this.amountCents,
    required this.currency,
    required this.source,
    required this.entryKind,
    required this.paymentMethod,
    this.notes,
    this.recordedBy,
    this.isDeleted = false,
    this.deletedAt,
    this.deletedBy,
    required this.createdAt,
    this.recordedByProfile,
    this.clientProfile,
    this.plan,
    this.gym,
    this.deletedByProfile,
  });

  // Getters for compatibility
  double get amount => amountCents / 100.0;
  Profile? get profile => clientProfile;
  DateTime get paymentDate => paidAt;
  MembershipPlan? get membershipPlan => plan;

  factory RevenueLedger.fromJson(Map<String, dynamic> json) {
    return RevenueLedger(
      id: json['id'] as String,
      paidAt: DateTime.parse(json['paid_at'] as String),
      planId: json['plan_id'] as String?,
      membershipId: json['membership_id'] as String?,
      gymId: json['gym_id'] as String?,
      amountCents: json['amount_cents'] as int,
      currency: json['currency'] as String,
      source: json['source'] as String,
      entryKind: json['entry_kind'] as String,
      paymentMethod: json['payment_method'] as String,
      notes: json['notes'] as String?,
      recordedBy: json['recorded_by'] as String?,
      isDeleted: json['is_deleted'] as bool? ?? false,
      deletedAt: json['deleted_at'] != null
          ? DateTime.parse(json['deleted_at'] as String)
          : null,
      deletedBy: json['deleted_by'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      recordedByProfile: json['recorded_by_profile'] != null
          ? Profile.fromJson(
              json['recorded_by_profile'] as Map<String, dynamic>,
            )
          : null,
      clientProfile: json['profile'] != null
          ? Profile.fromJson(json['profile'] as Map<String, dynamic>)
          : null,
      plan: json['membership_plan'] != null
          ? MembershipPlan.fromJson(
              json['membership_plan'] as Map<String, dynamic>,
            )
          : null,
      gym: json['gym'] != null
          ? Gym.fromJson(json['gym'] as Map<String, dynamic>)
          : null,
      deletedByProfile: json['deleted_by_profile'] != null
          ? Profile.fromJson(json['deleted_by_profile'] as Map<String, dynamic>)
          : null,
    );
  }

  String get amountInCurrency => '${amount.toStringAsFixed(2)} $currency';

  bool get isCharge => entryKind == 'charge';
  bool get isRefund => entryKind == 'refund';
}

File: kratos-admin-expo/src/models/revenue-ledger.ts
import { parseGym, type Gym } from '@/models/gym';
import { parseMembershipPlan, type MembershipPlan } from '@/models/membership-plan';
import { parseProfile, type Profile } from '@/models/profile';
import {
  asBoolean,
  requiredDate,
  requiredString,
  requiredInteger,
  asNullableDate,
  asNullableString,
  asRecord,
} from '@/utils/parsing';

export interface RevenueLedger {
  id: string;
  paidAt: Date;
  planId: string | null;
  membershipId: string | null;
  gymId: string | null;
  amountCents: number;
  currency: string;
  source: string;
  entryKind: string;
  paymentMethod: string;
  notes: string | null;
  recordedBy: string | null;
  isDeleted: boolean;
  deletedAt: Date | null;
  deletedBy: string | null;
  createdAt: Date;
  recordedByProfile: Profile | null;
  clientProfile: Profile | null;
  plan: MembershipPlan | null;
  gym: Gym | null;
  deletedByProfile: Profile | null;
}

export function parseRevenueLedger(value: unknown): RevenueLedger {
  const row = asRecord(value);
  return {
    id: requiredString(row.id),
    paidAt: requiredDate(row.paid_at),
    planId: asNullableString(row.plan_id),
    membershipId: asNullableString(row.membership_id),
    gymId: asNullableString(row.gym_id),
    amountCents: requiredInteger(row.amount_cents),
    currency: requiredString(row.currency),
    source: requiredString(row.source),
    entryKind: requiredString(row.entry_kind),
    paymentMethod: requiredString(row.payment_method),
    notes: asNullableString(row.notes),
    recordedBy: asNullableString(row.recorded_by),
    isDeleted: asBoolean(row.is_deleted),
    deletedAt: asNullableDate(row.deleted_at),
    deletedBy: asNullableString(row.deleted_by),
    createdAt: requiredDate(row.created_at),
    recordedByProfile: row.recorded_by_profile ? parseProfile(row.recorded_by_profile) : null,
    clientProfile: row.profile ? parseProfile(row.profile) : null,
    plan: row.membership_plan ? parseMembershipPlan(row.membership_plan) : null,
    gym: row.gym ? parseGym(row.gym) : null,
    deletedByProfile: row.deleted_by_profile ? parseProfile(row.deleted_by_profile) : null,
  };
}

export const revenueAmount = (entry: RevenueLedger) => entry.amountCents / 100;
export const revenueAmountLabel = (entry: RevenueLedger) =>
  `${revenueAmount(entry).toFixed(2)} ${entry.currency}`;
